File: src/main/java/j/Train.java
package j;

import k.Locomotive;
import k.RailroadCar;

import java.util.List;
import java.util.Objects;

@Deprecated
public class Train {

    private List<RailroadCar> railroadCars;
    private boolean running = false;

    public Train(final List<RailroadCar> composition) {
        Objects.requireNonNull(composition);
        railroadCars = composition;
    }

    public int getTotalWeight() {
        return railroadCars.stream().mapToInt(RailroadCar::getWeight).sum();
    }

    public int getTotalTowableMass() {
        return railroadCars.stream()
                .filter((c) -> c instanceof k.Locomotive)
                .map((c) -> (Locomotive) c)
                .mapToInt(Locomotive::getTowableMass)
                .sum();
    }

    public void start() {
        if (getTotalTowableMass() < getTotalWeight()) {
            throw new IllegalStateException("Too many coaches");
        }
        running = true;
    }

    public void stop() {
        running = false;
    }

    public boolean isRunning() {
        return running;
    }

    public List<RailroadCar> getRailroadCars() {
        return railroadCars;
    }
}

File: src/main/kotlin/k/Train.kt
package k

class Train(val railroadCars: List<RailroadCar>) {

    var running = false
        private set

    val totalWeight: Int
        get() = railroadCars.sumBy { it.weight }

    val totalTowableMass: Int
        get() {
            var total = 0
            for (car in railroadCars)
                if (car is Locomotive)
                    total += car.towableMass
            return total
        }
    //railroadCars
    // .filterIsInstance(Locomotive::class.java)
    // .sumBy { it.towableMass }

    fun start() {
        check(totalTowableMass >= totalWeight) { "Too many coaches" }
        running  = true
    }

    fun stop() { running  = false }

    override fun toString(): String = railroadCars.joinToString(".")
}
